File: semester-2/lab-5/task-2/C++/BaseString.cpp
//
// Created by Arthur Ladchenko on 15.04.2020.
//

#include <cstring>
#include "BaseString.h"

int BaseString::GetLength() {
    return strlen(this->value);
}

char *BaseString::GetValue() {
    return this->value;
}

BaseString::BaseString(char *value) {
    this->value = value;
}

int BaseString::CountChar(char c) {
    int counter = 0;
    for (int i = 0; i < this->GetLength(); ++i) {
        if (this->value[i] == c) counter++;
    }
    return counter;
}

void BaseString::RemoveCharacter(char c) {
    char* newString = new char[this->GetLength() - this->CountChar(c)];
    int step = 0;
    for (int i = 0; i < this->GetLength(); i++) {
        if (c == this->value[i]) { continue; }
        newString[step] = this->value[i];
        step++;
    }
    value = newString;
}

File: semester-2/lab-5/task-2/C#/BaseString.cs
using System.IO;
using System.Linq;

namespace lab_5_2
{
    public class BaseString
    {
        protected char[] StrValue { get; set; }

        public BaseString(params char[] str)
        {
            StrValue = str;
        }

        public int GetLength() => StrValue.Length;

        public char[] GetValue() => StrValue;

        public virtual void RemoveCharacter(char l)
        {
            char[] newValue = new char[GetLength() - StrValue.Count(c => c == l)];
            int step = 0;
            if (StrValue.Contains(l))
            {
                for (int i = 0; i < GetLength(); i++)
                {
                    if (l == StrValue[i]) {continue;}
                    newValue[step] = StrValue[i];
                    step++;
                }

                StrValue = newValue;
            }
            else
            {
                throw new InvalidDataException("Char " + l + " does not exist in MyString object.");
            }
        }
    }
}